kong/clustering/wrpc_control_plane.lua

```diff
@@ -0,0 +1,662 @@
+      end
+    end)
+
+    wrpc_config_service:set_handler("ConfigService.ReportMetadata", function(peer, data)
+      local client = self.clients[peer.conn]
+      if client then
+        ngx_log(ngx_INFO, _log_prefix, "received initial metadata package from client: ", client.dp_id)
+        client.basic_info = data
+        client.basic_info_semaphore:post()
+      end
+      return {
+        ok = "done",
+      }
+    end)
+  end
+
+  return wrpc_config_service
+end
+
+
+local function extract_major_minor(version)
+  if type(version) ~= "string" then
+    return nil, nil
+  end
+
+  local major, minor = version:match(MAJOR_MINOR_PATTERN)
+  if not major then
+    return nil, nil
+  end
+
+  major = tonumber(major, 10)
+  minor = tonumber(minor, 10)
+
+  return major, minor
+end
+
+
+local function plugins_list_to_map(plugins_list)
+  local versions = {}
+  for _, plugin in ipairs(plugins_list) do
+    local name = plugin.name
+    local version = plugin.version
+    local major, minor = extract_major_minor(plugin.version)
+
+    if major and minor then
+      versions[name] = {
+        major   = major,
+        minor   = minor,
+        version = version,
+      }
+
+    else
+      versions[name] = {}
+    end
+  end
+  return versions
+end
+
+
+function _M.new(parent)
+  local self = {
+    clients = setmetatable({}, { __mode = "k", }),
+    plugins_map = {},
+  }
+
+  return setmetatable(self, {
+    __index = function(tab, key)
+      return _M[key] or parent[key]
+    end,
+  })
+end
+
+
+local config_version = 0
+
+function _M:export_deflated_reconfigure_payload()
+  local config_table, err = declarative.export_config_proto()
+  if not config_table then
+    return nil, err
+  end
+
+  -- update plugins map
+  self.plugins_configured = {}
+  if config_table.plugins then
+    for _, plugin in pairs(config_table.plugins) do
+      self.plugins_configured[plugin.name] = true
+    end
+  end
+
+  config_version = config_version + 1
+
+  -- store serialized plugins map for troubleshooting purposes
+  local shm_key_name = "clustering:cp_plugins_configured:worker_" .. ngx.worker.id()
+  kong_dict:set(shm_key_name, cjson_encode(self.plugins_configured));
+
+  local service = get_config_service(self)
+  self.config_call_rpc, self.config_call_args = assert(service:encode_args("ConfigService.SyncConfig", {
+    config = config_table,
+    version = config_version,
+  }))
+
+  return config_table, nil
+end
+
+
+function _M:push_config()
+  local payload, err = self:export_deflated_reconfigure_payload()
+  if not payload then
+    ngx_log(ngx_ERR, _log_prefix, "unable to export config from database: ", err)
+    return
+  end
+
+  local n = 0
+  for _, client in pairs(self.clients) do
+    client.peer:send_encoded_call(self.config_call_rpc, self.config_call_args)
+
+    n = n + 1
+  end
+
+  ngx_log(ngx_NOTICE, _log_prefix, "config version #", config_version, " pushed to ", n, " clients")
```

Comment:
I notice that in the non-wrpc version, we log in a different format. Should we unify them?

Reply:
Also, the log level is different.